doc/design/state-machine.md

```diff
@@ -0,0 +1,185 @@
+
+Operation: none.
+
+__e<sub>5</sub>__:
+
+Condition: a Reserving or Reserved cell in this AG is being overwritten (e<sub>3</sub> or e<sub>6</sub> in cell state machine).
+
+Operation: set all the other Reserving cells to Used (e<sub>4</sub> in cell state machine), Reserved cells to Free (e<sub>7</sub> in cell state machine).
+
+__e<sub>6</sub>__:
+
+Condition: a cell from Used to Reserving (e<sub>2</sub> in cell state machine)
+
+Operation: none.
+
+__e<sub>7</sub>__:
+
+Condition: AG deleted.
+
+Operation: set the Reserved cells to Free (e<sub>7</sub> in cell state machine), set the Reserving cells to Used (e<sub>4</sub> in cell state machine).
+
+## Cell State Machine
+
+Cell is the resource unit in HiveD. The figure below shows the state machine of cell. Note that here cells are _lowest-level physical cells_ (we record states only in these cells).
+
+![AG](cell-state-machine.png)
+
+### States
+
+__Free__: the cell is not allocated to any AG.
+
+__Used__: the cell is allocated to an AG, and the AG is running on this cell.
+
+__Reserved__: the cell is allocated to an AG, but there is no AG running on this cell (e.g., the AG is Preempting).
+
+__Reserving__: the cell is allocated to an AG, but another AG is still running on the cell (e.g., the first AG is Preempting the running AG).
+
+Note that the reservation of cells (Reserved and Reserving states) is not necessarily designed for preemptions (i.e., reserving resources for the Preempting AGs). In the future it is possible that we extend this mechanism to support other reservation mechanisms.
+
+Common cell life cycles:
+
+Free -> Used -> Free (no preemption involved)
+
+Free -> Used -> Reserving -> Reserved -> Used -> Free (preemption involved)
+
+### Events
+
+Note: “cell view” in the below operations represents the in-memory data structures for scheduling, e.g., free cell list, cell bindings, cell priorities. Allocating / releasing a cell in the cell view will modify the free cell list (and split / merge the cell), create / destroy cell bindings, and set / reset the cell priority. These changes are immediately visible to the subsequent scheduling cycles.
+
+__e<sub>c</sub>__:
+
+Condition: scheduler crashes and restarts.
+
+Operation: none.
+
+Note that all states are volatile; the Free and Used states are derived from the AG state machine.
+
+__e<sub>0</sub>__:
+
+Condition: triggered by AG from Pending to Allocated (e<sub>0</sub> in AG state machine).
+
+Operation: allocate the cell to the AG in the cell view.
+
+__e<sub>1</sub>__:
+
+Condition: the Allocated AG is deleted (e<sub>3</sub> in AG state machine).
+
+Operation: release the cell in the cell view.
+
+__e<sub>2</sub>__:
+
+Condition: triggered by AG from Pending to Preempting (e<sub>1</sub> in AG state machine).
+
+Operation: set the Allocated AG on this cell to Being preempted state (e<sub>6</sub> in AG state machine); release the cell from the cell view, and then allocate it to the preemptor AG in the cell view.
```

Comment:
Can we use markdown code mark for the states in doc, such as 
Being preempted
->
`Being preempted`